### ISSUE TITLE
improve docs

### ISSUE THREAD

**Issue by tylerflex** [flexcompute/tidy3d#1148]
Some things we could do to make this better:

### Landing page
- [x] Toggle between command line and python instructions for installation and configure.
- [x] (possibly?) Toggle between linux / mac and windows instructions
- [ ] Add link to colab notebook with the minimal example and API key loaded as environment variable.
- [X] Add nice banners

### Aesthetics
- [x] Improve overall theme.
- [X] Make the font better
- [X] Improve general page styling so it's more readable
- [x] Add emojis, generally make the docs look nicer.
- [X] wider columns for titles and code.
- [x] make table of contents display more logical
- [x] Find a dark theme logo that works - I think important because I and I assume others, exclusively work in dark themes. Currently chatting with Qing. (I can modify the existing one easily if approved)
- [ ] Cooler intro page and pictures
- [x] Make toctrees nicer

### API Documentation Improvements
From @e-g-melo As an user of documentation, I think that it would help a lot if 
- [ ] The user can find more context when reading though the class definitions 
- [ ] Including equations when needed 
- [ ] Schematics to show geometric parameters
- [ ] An example on how to use that object in a simulation.
- [ ] Link to examples and tutorials
- [x] We need to sort out the way the attributes are propagated.
- [ ] Cleaner API viewer
- [X] Solve the hierarchical class method search issue https://github.com/flexcompute/tidy3d/issues/1197
- [X] Hide irrelevant private methods on API
- [X] ~Sort out the material library toctree situation~ TODO autosummary generate restructure in this case re fundamental hierarchy
- [X] Improve the API tab search and hierarchy
- [ ] Make nice illustration of medium property types, ie anisotropic isotropic for the uninitated
- [ ] Make nice illustrations of all the sources and geometrically how they are represented
- [ ] Standalone examples for most common class methods etc.
- [x] Solve the class page index issue on the right, especially for Simulation that it became quite big.
- [X] Any other way to solve the inheritance method propagation issue apart from the manual autosummary? Maybe there's no alternative and this is it. Completeness by documenting all base classes on absract/ done
- [x] Fix the m2r2 situation. This may require a patch of that repository.
- [x] Backup a main and develop branch of the `tidy3d-docs` just in case which might get overwritten
- [ ] Potentially work on the adjoint docstrings depending on what Tyler wants re his restructure
- [x] Fix all the broken links from the 2.5 structure

### Functionality
- [x] how to build locally.
- [ ] remove all warnings.
- [ ] Build pdf automatically receipe
- [X] Docstring updating script so it's easier to change APIs throughout the multiple versions

### New Structure
- [x] Write down poetry instructions
- [x] Document development `tidy3d` scripts
- [X] Create the ported `tidy3d-notebooks`. Update already one exists, bring it back to life?

### Pre-Integration Checklist
- [x] Make sure all the github action workflows pass after the upgrade
- [x] Verify tests
- [x] Think about how to port the latest docs repo changes to be integrated? Mantain history or how do we do this?
- [x] How do we deal with versioning here?
- [x] Sort out the final home of the notebooks repo, the mirror repo, and the main source code repo
- [x] Add docstring image example Emerson suggested and sent

### Implementation Tasks
- [x] Upgrade the sphinx version. I believe a lot of what we want out of documentation can be solved in a first pass through a theme dependency upgrade (because I’ve used the same theme before, haha - so I know!) There's some packages conflicts I'm currently dealing with.
- [x] Guarantee important equivalence between the new and old documentation `conf.py`
- [x] Demonstrate reproducible working environment using `poetry`
- [ ] ~Remove old unnecessary dependencies~ I feel this should be done by someone deeply involved in the project.
- [x] Sort out this table of contents situation
- [x] Write documentation verification environment script on `tidy3d cli`
- [x] Write documentation build script on `tidy3d cli`
- [x] Write automatic configuration of notebook submodule on `tidy3d cli`
- [x] ~Do we like [this live code binder integration](https://sphinx-book-theme.readthedocs.io/en/stable/reference/thebe.html) of our current theme. I'm kind of keen~
- [x] Setup the colab notebooks and live notebook links on the selected notebooks we want.
- [x] Install a few pre-commits to guarantee docs + code quality.
- [x] Commit between the two repositories
- [X] Restructure the docs project within the project
- [x] Remove bugs from the installation scripts
- [x] Verify reproducibility of the envrionment installation scripts
- [x] Thoroughly verify that the documentation contains all the same information as before.
- [ ] Update README.md for notebooks
- [x] Update README.md for source
- [x] Reorganise the image img paths from the restructure
- [x] Sort out the sitemap building issue. Why is it failing? Last equivalence point I think.
- [x] Do we need to set up a version branch deployment action for RTD?
- [x] Have I screwed up the `draw_package` and `generate_doc` in `docs`?
- [x] FIX all the API links in the notebooks
- [x] Improve the way that images and equations get rendered into our API theme.
- [x] Fix dark pygments viewing on dark mode in web outputs.

### Notebook repo:
- [X] Precommit sorting out `.ipynb` notebooks conflicts automatically

## Implementation Plan
- [X] Merge all the latest changes from `tidy3d-docs` into `tidy3d-notebooks`.
- [X] Update the new structure `tidy3d` submodule links to this develop branch.
- [X] Set up live demo project before actual merge that demonstrates live development workflow.
- [x] Fix all the broken API links in the notebooks
- [X] Update the readthedocs deployment branches to develop, and verify the github actions should work for those branches too
- [x] Update the latest `develop` branch of `tidy3d-notebooks` accordingly the day we decide we won't use `tidy3d-docs` anymore.
- [x] DANGER: Change the docs mirror repository name on the `sync-readthedocs-repo` action on the `tidy3d` develop branch. This will overwrite the `develop` branch so that it is mirroring the new `tidy3d` structure.
- [x] Verify that the new notebooks repository branch has all the github actions that still work for web deployment.
- [X] Merge the `repo_merge_no_history` branch of the `tidy3d` repo into the `develop` branch.
- [x] Bring back python 3.8 support

## Post First Implementation Plan

- [ ] Upgrade the jax requirements to caret and test across all platforms
- [ ] Test tidy3d develop improvements to windows installation
- [ ] Explore pdf building
- [x] Restructure back to python3.8
- [x] Upgrade the github actions accordingly.
- [x] Whether the backend still works with this commit. Can confirm with momchil, daniil, or weiliang.
- [x] Whether the SEO works as expected with the newly hosted docs page. maybe you can ping yongwei in the em_all channel.
- [x] Finish migrating all the pre/2.5 docs
- [x] Finish migrating all the pre/2.5 notebooks
- [x] Sort out the Changelog.md situation
- [ ] CLI test single notebook via arguments
- [ ] Update the development docs to the latest flow.
- [x] Fix pandas table visualisation
- [ ] See if we can remove the extra API classes nder the table
- [ ] Include the relevant markdown files inot their corresponding API sections.
- [x] Integrate the FAQ from the new repo

Final deployment tasks:
- [ ] Replace all hardcoded links with relative links in order to avoid version mismatches in docs
- [x] Replace all hardcoded links with absolute links in order to avoid version mismatches in notebooks
- [x] Update 404 links to warn of mismatched paths
- [X] Write tests to check for latest submodule
- [X] FAQ submodule

**Comment by momchil-flex** [comment]
Just wanted to +1 improving the table of contents organization as maybe the most important thing in my mind. As far as I remember we tried to play around with this a bit but the sphinx template was hard to understand (we *could* consider a different template..) But basically there is no organization in the left panel, which makes it almost useless for our example library and our api reference

![image](https://github.com/flexcompute/tidy3d/assets/92756559/89379b1d-927b-4ef0-a9dd-38d7dba7e553)

I sometimes use the contents on the right, but it's not as intuitive, and if you're looking on mobile you don't even see that.

I'm bringing this up not only because it has been bugging me personally, but I had the same complaint from a user yesterday - that our examples are not organized. I showed him the main page and he was like oh ok, but he had been looking at the left panel up to then.

**Comment by daquintero** [comment]
Hi guys, all sounds good and I also agree with those points - cheers for the clear plan ahead too! I'm already working on this on these branches and will PR when several points are solved.

https://github.com/daquintero/tidy3d/tree/docs_improvement_suggestions
https://github.com/daquintero/tidy3d-docs/tree/docs_improvements_suggestions

**Comment by tylerflex** [comment]
Thanks @daquintero ! There are also some related issues tagged "daq" in the front end that I was planning to consolidate into here at some point, but if you are already working on it, might want to check those out too.

I think we can wait for your first day (Wednesday?) as then you get onboarding and will get a flexcompute email address to use.

**Comment by daquintero** [comment]
Yeah, I was checking them out to be more familiar when starting! Sounds good, we'll chat then

**Comment by daquintero** [comment]
So I've been thoroughly looking at a way that solves many of the issues in a single shot, and a strategy to propose. Mainly to get up to speed to contribute well when I start.

In summary, based on the analysis of this "brainstorming notes" discussion https://github.com/flexcompute/tidy3d/discussions/1220  (please you don't need to read it all, it's more like a scratch paper to me):

> I think a good approach is this: we have a `pyproject.toml` that controls all the requirements for the main project, and documentation + tests. This means that we have full control over how to create a virtual environment that will be functionally correct and that is shared and verified between the two source and documentation projects. We can build a mambaconfiguration  installs script eg. [for automatic environment configuration eg. with binaries such as `pandoc`](https://medium.com/hydroinformatics/how-to-make-your-python-environment-reproducible-common-practices-conda-environment-de28195b74de) and for [reproducible packaging recipe](https://docs.conda.io/projects/conda-build/en/stable/concepts/recipe.html) for multiple platforms based on this. Also the mamba `pandoc` mamba recipe already includes the actual binary which apparently was an issue when reproducing the local install in the documentation repository. 

This might be an integrated solution for improving an easier local build in https://github.com/flexcompute/tidy3d/issues/1148 ,  having mantainable and distributable multi-platform environment and package distributed via mamba, and performs an upgrade of the current setup.py scripts  too re https://github.com/flexcompute/tidy3d/issues/1068 .

Maybe we'll chat more when I start, for now, I am just getting to speed so won't implement this until we chat on Wednesday.

**Comment by daquintero** [comment]
Just saw your message on the discussions and thanks for catching me up to speed, much appreciated! Really interesting chat and thinking about it

**Comment by tylerflex** [comment]
No problem, great to be having these discussions. I think it would be useful for you to continue taking a fresh perspective on this to come up with an approach to organizing things that makes sense and simplifies our development process. The only real "constraints" are:
1. we don't want huge notebook diffs to clutter our git history.
2. the docs must be hosted by `flexcompute-readthedocs` organization (not `flexcompute`), unless we can find a way around this to restrict `readthedocs` from seeing our private repos.

and in general I think the main "nice to have"s from my perspective (in addition to the minor things scattered around in the issues):
A. simplify the workflow of adding a new feature, potentially requiring (a) backend changes (b) frontend changes, and (c) a new notebook or some additions to the docs. Ideally, we'd like (b) and (c) to be combined and tracked in the same PR.
B. A single `pyproject.toml` containing all of the package requirements with both flexibility for dependency grouping (docs, features, etc) but also requirement version dependencies being defined in an organized way that avoids issues (eg. using poetry).

**Comment by Frank-DingYC** [comment]
@tylerflex These days, I have read the code on sources. I found there were no documentation about the source_time. I make some try on these part. Here are the markdown file and source code on these part.
Hopefully, these can help reader to understand the source waveform well~
[Source_time.zip](https://github.com/flexcompute/tidy3d/files/13329886/Source_time.zip)

**Comment by daquinteroflex** [comment]
Just writing down the changes in progres:

- [ ] Upgrade the sphinx version. I believe a lot of what we want out of documentation can be solved in a first pass through a theme dependency upgrade (because I’ve used the same theme before, haha - so I know!) There's some packages conflicts I'm currently dealing with.

The good news is that I've been testing the new `pyproject.toml` environment and am pretty happy using it.

**Comment by tylerflex** [comment]
@Frank-DingYC 

Thanks for contributing the explanation on the various types of source time.

A few notes:

> I found there were no documentation about the source_time.

The [GaussianPulse](https://docs.flexcompute.com/projects/tidy3d/en/latest/_autosummary/tidy3d.GaussianPulse.html?highlight=GaussianPulse) is present in the documentation.

We just recently added the other two officially to the API in this [commit](https://github.com/flexcompute-readthedocs/tidy3d-docs/pull/308/files), but it hasn't been released yet, so those will come out next release.

Note we also have an issue open #1243 for adding the equations into the documentation, which will help out.

I think your explanation and the plots are useful, I'm just not really sure how to include them right now.  Maybe we could put a section in the FAQ explaining what kind of source times are available and use your plots? I don't think the python notebook is extensive enough to warrant its own separate example yet. @tomflexcompute do you have any thoughts on this?

**Comment by tylerflex** [comment]
@daquintero 


> * [ ]  Upgrade the sphinx version. I believe a lot of what we want out of documentation can be solved in a first pass through a theme dependency upgrade (because I’ve used the same theme before, haha - so I know!) There's some packages conflicts I'm currently dealing with.
> 
> The good news is that I've been testing the new `pyproject.toml` environment and am pretty happy using it.

That's great. Yea we've been stuck with an older versions because nobody could figure out a way out of the dependency hell. Glad to hear that the new setup could be helping a lot in this regard!

**Comment by tomflexcompute** [comment]
> @Frank-DingYC
> 
> Thanks for contributing the explanation on the various types of source time.
> 
> A few notes:
> 
> > I found there were no documentation about the source_time.
> 
> The [GaussianPulse](https://docs.flexcompute.com/projects/tidy3d/en/latest/_autosummary/tidy3d.GaussianPulse.html?highlight=GaussianPulse) is present in the documentation.
> 
> We just recently added the other two officially to the API in this [commit](https://github.com/flexcompute-readthedocs/tidy3d-docs/pull/308/files), but it hasn't been released yet, so those will come out next release.
> 
> Note we also have an issue open #1243 for adding the equations into the documentation, which will help out.
> 
> I think your explanation and the plots are useful, I'm just not really sure how to include them right now. Maybe we could put a section in the FAQ explaining what kind of source times are available and use your plots? I don't think the python notebook is extensive enough to warrant its own separate example yet. @tomflexcompute do you have any thoughts on this?

Thanks. A separate example would probably be too basic. FAQ entry sounds good. I think eventually it would be nice to have capabilities in the API references to show schematic pictures. Currently we show equations in a few places but they can be added to more.

**Comment by daquinteroflex** [comment]
I must say I've only just come to appreciate how much of a dependency hell it actually is! Cracking on with a good strategy I think - a lot of the issues I think are related to reproducible environment toolset upgrades

**Comment by tylerflex** [comment]
@Frank-DingYC 

Maybe you could try making a pull request into tidy3d-docs adding a short discussion on the types of source time?

Basically this will involve an edit of this file https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/develop/docs/source/faq.rst

You can see in the file some examples of how we do links and images in `rst` and it could be a good, simple exercise to learn how to contribute in the future.

**Comment by Frank-DingYC** [comment]
OK~ I wil make some try on this part~

**Comment by daquinteroflex** [comment]
Unsure if I am running into a new issue here, but I am finding that if I have more than a few notebooks open my Pycharm IDE really struggles with the cache and crashes (and I would say it's a top end computer). Maybe users are running into this and we  wouldn't know, but maybe my case is an edge case.

This relates to what @lucas-flexcompute mentioned in [this dicusssion](https://github.com/flexcompute/tidy3d/discussions/1220):

>  was also looking into jupytext when trying to improve our workflow, but I could not find ans easy way to have the run results also cached in the repo. The discussion I had with Tyler and Momchil at the time was that having the cell outputs in the repository is very desirable, because PR reviews are much easier when we can compare plots side-by-side, for example, straight in github. It's also convenient to have the notebook results ready when we're making any changes without having to run them before we start (some notebooks can be very long).

> My ideal solution would be to have the jupytext version of each notebook in the repository history and the ipynb versions only for the branch tip, but I don't know if that's viable. If jupytext and ipynb are kept in sync, we might be able to set up a git hook that removes old versions of the ipynb when we commit, but both working with synced files and using commit hooks are prone to error.

I I believe it would be somewhat valuable to implement something on the lines of this idea, as some of these files are massive. But maybe there's no workaround at wanting to save run history and they have to be large. Am just thinking how it fits in the current flow, and in any case we can do so later on after the first merge.

**Comment by daquinteroflex** [comment]
So there's a minor issue I've been trying to solve, and just putting feelers out there because the solution might involve code restructuring, or just leaving it as is:

See the new [Simulation class](https://flexcompute-tidy3d-docs-demo.readthedocs-hosted.com/en/repo_merge_no_history/api/_autosummary/tidy3d.Simulation.html#tidy3d.Simulation) and scroll to the bottom. You  might notice that the properties, attributes and methods for the generated classes are a bit all over the place as you scroll down. This is because sphinx generates the documentation in the order in which the code is written [PEP 520 ](https://peps.python.org/pep-0520/).

The solution might involve restructuring the classes so that things appear in order in the documentation, which is something maybe there is could be some hesitancy in introducing unknown bugs. Other people have evaluated this problem and it's mention it's very difficult to change from a pure sphinx side. Info in related issues: https://github.com/software-mansion/starknet.py/issues/716 and https://github.com/sphinx-doc/sphinx/issues/5270

**Comment by tylerflex** [comment]
I dont foresee see too much of a problem with this, I'd say it's ok to reshuffle the methods.  The only thing we need to be aware of is that I think the order is important for pydantic validators. For example, if validator A modifies the value somehow, then it goes to validator B. If we reverse the order, there could be effects there. But I think the methods and properties can be moved around or regrouped as needed.

**Comment by daquinteroflex** [comment]
Edits in progress:

Just merged into 2.5 and writing up here a list of changes:

* `MANIFEST.in` becomes a part of [include and exclude poetry statements](https://python-poetry.org/docs/pyproject/#include-and-exclude). File gets removed
* `create_config_folder` in `setup.py` gets moved to the `__init__` declaration as `pyproject.tomls` are declarative, not functional. I can hack it but does this move make sense anyway? Unsure about it since it includes an os import on __init__.py, maybe we need to discuss more. `setup.py` gets replaced by `pyproject.toml`
* Important to note I had replaced `tox.ini` with a custom script that uses poetry on github actions when I was first debugging poetry - which I think is equivalent. It has been updated with `vtk` and I have updated it with the `vtk` commands now. Do we want to matain local testing with tox? I can see a workaround to test it with poetry too both on GhActions and locally if we want. @momchil-flex

**Comment by daquinteroflex** [comment]
So we just need to take the decision above pending rebase.There are many things to fix still but generally can be rebased imo so we can begin merging and updating the simulation flow including the docs live update.

**Comment by tylerflex** [comment]
I'll look in more detail soon but regarding 3.8 support: Is there a way perhaps to support >= 3.8 generally but only >= 3.9 for the adjoint plugin (jax dependency group) and windows users? Eg if a windows user tries to pip install tidy3d[jax] with python < 3.9 we provide a useful error message?

**Comment by daquinteroflex** [comment]
I think we can do that, the problem will be that the `jax` tests won't pass for that version. We can also try to create test groups which we maybe have to do anyway for the notebooks as it's all consolidated. This just means it's an upgrade on the tests pipeline accordingly. 

Are you also able to help me when you can with configuring the rebase properly? I think I accidentally merged due to the automatic git config, rather than actually rebase. Going to backup in a new branch but unsure how to best approach this, or start from a new branch instead and merge again all the changes? Not inclined to keep adding new features until I'm properly set up to the flow.

**Comment by tylerflex** [comment]
Here's one thing I noticed, the extra elements in the TOC at the bottom are kind of distracting.
<img width="335" alt="image" src="https://github.com/flexcompute/tidy3d/assets/92756888/bcd26ac3-a5c2-4e78-8191-0ccacfdcb9d8">
should they removed or put under their own header?

**Comment by tylerflex** [comment]
Another issue: pandas DataFrames dont render very well in this dark mode. Is there a way around this?:

<img width="847" alt="image" src="https://github.com/flexcompute/tidy3d/assets/92756888/228af021-9341-4b21-bbe6-b7b67a5defd5">

**Comment by tylerflex** [comment]
what's the status with this issue? still in progress?

**Comment by daquinteroflex** [comment]
Yeah so I'm editing the tasks at the top to consolidate the missing things

**Comment by daquinteroflex** [comment]
Closing this in favour of the post migration https://github.com/flexcompute/tidy3d/issues/1568
